Export multi-tabbed grids to excel

Sigmonia opened the pull request:
#### Rationale
Item #10158: Multi-sheet export to Excel workbooks of multi-tab grids

#### Related Pull Requests
* [platform](https://github.com/LabKey/platform/pull/3273)
* [samplemanagement](https://github.com/LabKey/sampleManagement/pull/926)
* [biologics](https://github.com/LabKey/biologics/pull/1262)
* [inventory](https://github.com/LabKey/inventory/pull/422)
* [js-api](https://github.com/LabKey/labkey-api-js/pull/127) -- To allow reading of download errors

#### Changes
* Add selection modal to tabbed grids to enable selecting from available tabs

Sigmonia committed:
add modal to tabbed grids to select tabs for export

Sigmonia committed:
Bump alpha package version and draft release notes

Sigmonia committed:
slight change to draft release notes

Sigmonia committed:
add exportFilename property to tabbed grids

Sigmonia committed:
v2.155.1-fb-excelTabs.1

Sigmonia committed:
merge from develop

Sigmonia committed:
v2.156.1-fb-excelTabs.2

Sigmonia committed:
Adjust filename used when exporting tabbed sample grids

Sigmonia committed:
v2.156.1-fb-excelTabs.3

labkey-matthewb commented on packages/components/src/internal/components/gridbar/ExportModal.tsx:
```diff
@@ -0,0 +1,74 @@
+    return (<>
+        <Modal onHide={closeHandler} show={true}>
+            <Modal.Header closeButton>
+                <Modal.Title>Select the Tabs to Export to Excel</Modal.Title>
```
to "Excel Workbook" ?

cnathe replied:
This `ExportModal` seems generic and could be used for other file types in the future, so I might suggest that this `Excel` specific part be passed in as a `title` prop.

Sigmonia replied:
added

Sigmonia committed:
Update export tests

Sigmonia committed:
merge from develop

Sigmonia committed:
v2.157.0-fb-excelTabs.4

cnathe commented on packages/components/src/internal/actions.ts:
```diff
@@ -535,6 +535,32 @@ export function getExportParams(
+    return new Promise((resolve, reject) => {
+        const controller = 'query';
+        const action = 'exportTabsXlsx.api';
+        const url = ActionURL.buildURL(controller, action, undefined);
```
is this third param for `undefined` needed?

Sigmonia replied:
no, removed.

cnathe commented on packages/components/src/internal/components/gridbar/ExportModal.tsx:
```diff
@@ -0,0 +1,74 @@
+    queryModels: QueryModelMap;
+    tabOrder: string[];
+    onExport: (tabs: Set<string>) => Promise<void>;
+    onClose?: () => void;
```
it seems like `onClose` should be a required prop for a modal component

Sigmonia replied:
done

cnathe commented on packages/components/src/internal/components/gridbar/ExportModal.tsx:
```diff
@@ -0,0 +1,74 @@
+            setSelected(new Set(selected.add(modelId)));
+        }
+        else {
+            selected.delete(modelId);
```
does this delete act on the `selected` variable directly? do we want to instead create a new set here and then delete from that set and call `setSelected(updatedSelected)` or something along those lines?

Sigmonia replied:
done

cnathe commented on packages/components/src/internal/components/gridbar/ExportModal.tsx:
```diff
@@ -0,0 +1,74 @@
+                <Modal.Title>Select the Tabs to Export to Excel</Modal.Title>
+            </Modal.Header>
+            <Modal.Body>
+                <div className="excel-export-modal-body" >
```
similar comment as above, I'd suggest taking the word `excel` out of here since the rest of the component is generic.

Sigmonia replied:
done

cnathe commented on packages/components/src/internal/components/picklist/PicklistListing.tsx:
```diff
@@ -37,6 +37,7 @@ interface OwnProps {
     user: User;
     initTab?: string;
     CreateButton?: ComponentType<PicklistCreateButtonProps>;
+    picklistName?: string;
```
I don't see this new prop used anywhere. Is it needed?

Sigmonia replied:
removed

cnathe commented on packages/components/src/internal/components/samples/SampleAssayDetail.tsx:
```diff
@@ -306,6 +306,7 @@ export const SampleAssayDetailBodyImpl: FC<SampleAssayDetailBodyProps & Injected
             onTabSelect={onTabChange}
             activeModelId={activeTabId}
             title="Assay Results"
+            exportFilename={'Assay_Samples'} //TODO find name of assay
```
is this a TODO for this PR or a future PR?

cnathe commented on packages/components/src/public/QueryModel/TabbedGridPanel.tsx:
```diff
@@ -114,6 +126,42 @@ export const TabbedGridPanel: FC<TabbedGridPanelProps & InjectedQueryModels> = m
+            return;
+        }
+
+        exportTabs([internalActiveId]).then();
```
Since `exportTabs` uses async/await, I don't think the return type needs to be a `Promise`. If that return type is changed to just `void` then you should be able to remove the `then()` call here.

Sigmonia replied:
removed, seems like it still wants the promise

cnathe commented on packages/components/src/public/QueryModel/TabbedGridPanel.tsx:
```diff
@@ -114,6 +126,42 @@ export const TabbedGridPanel: FC<TabbedGridPanelProps & InjectedQueryModels> = m
+        }
+    }, [exportFilename, canExport, queryModels]);
+
+    const exportHandler = useCallback(async () => {
```
I was a bit confused at first thinking that all export types would call this callback, but it is just excel (based on the definition of `exportHandlers` at line 159. I might suggest renaming this to `excelExportHandler` just to self document.

Sigmonia replied:
done

Sigmonia committed:
Bind query parameters versus using custom query form
Merge QueryView.AliasColumnExcelWriter --> ExcelWriter
Fix renamed columns not being used
Fix query filters not being applied

Sigmonia committed:
merge from develop

Sigmonia committed:
merge from develop

Sigmonia committed:
Merge remote-tracking branch 'origin/develop' into fb_excelTabs

labkey-nicka commented on packages/components/src/internal/components/gridbar/ExportModal.tsx:
```diff
@@ -0,0 +1,75 @@
+    const [selected, setSelected] = useState<Set<string>>(new Set(tabOrder));
+
+    const closeHandler = useCallback(() => {
+        onClose?.();
```
`onClose` is required by the interface definition. You can remove this `?` and fix up any tests to actually include an `onClose`.

labkey-nicka commented on packages/components/src/internal/components/gridbar/ExportModal.tsx:
```diff
@@ -0,0 +1,75 @@
+    }, [onClose]);
+
+    const exportHandler = useCallback(() => {
+        onExport?.(selected);
```
`onExport` is required by the interface definition. You can remove this `?` and fix up any tests to actually include an `onExport`.

labkey-nicka commented on packages/components/src/internal/components/gridbar/ExportModal.tsx:
```diff
@@ -0,0 +1,75 @@
+    }, [selected]);
+
+    if (queryModels == null)
+        return undefined;
```
nit: It is preferred in React to return `null` for component hierarchies.

labkey-nicka commented on packages/components/src/internal/components/gridbar/ExportModal.tsx:
```diff
@@ -0,0 +1,75 @@
+    if (queryModels == null)
+        return undefined;
+
+    return (<>
```
nit: There is no need for this extra `<>` fragment wrapping the `Modal`.

labkey-nicka commented on packages/components/src/internal/components/samples/SampleAssayDetail.tsx:
```diff
@@ -306,6 +306,8 @@ export const SampleAssayDetailBodyImpl: FC<SampleAssayDetailBodyProps & Injected
             onTabSelect={onTabChange}
             activeModelId={activeTabId}
             title="Assay Results"
+            // exportFilename={sampleModel.name + '_assay_results'}
```
nit: Remove comment.

labkey-nicka commented on packages/components/src/public/QueryModel/TabbedGridPanel.tsx:
```diff
@@ -147,9 +200,24 @@ export const TabbedGridPanel: FC<TabbedGridPanelProps & InjectedQueryModels> = m
+                    {...rest}
+                />
             </div>
+            {queryModels !=null && showExportModal &&
```
nit: Try to avoid using `==` and `!=`. I would recommend `{showExportModal && !!queryModels ...`.

labkey-nicka commented on packages/components/src/internal/components/gridbar/ExportModal.tsx:
```diff
@@ -0,0 +1,75 @@
+
+export const ExportModal: FC<ExportModalProperties> = memo((props) => {
+    const {queryModels, tabOrder, onClose, onExport, canExport, title = DEFAULT_TITLE} = props;
+    const [selected, setSelected] = useState<Set<string>>(new Set(tabOrder));
```
nit: Consider using the function initializer to avoid instantiating the new set of subsequent renders.

```tsx
const [selected, setSelected] = useState<Set<string>>(() => new Set(tabOrder));
```

Sigmonia committed:
Update export file name for SampleAssayDetail page

labkey-nicka commented on packages/components/src/internal/actions.ts:
```diff
@@ -535,6 +535,32 @@ export function getExportParams(
     return params;
 }
 
+export async function exportTabsXlsx(filename:string, models: SchemaQuery[]): Promise<void>
```
A few comments.

1. This method doesn't need to be marked as `async` as it it creating a new `Promise` and returning it explicitly.
2. For failures at the request level we tend to do a `console.error()`.
3. Consider using `Utils.getCallbackWrapper()` and  `resolveErrorMessage()` rather than parsing the error message yourself.

Here is my recommendation:

```tsx
export function exportTabsXlsx(filename: string, tabModels: SchemaQuery[]): Promise<void> {
    return new Promise((resolve, reject) => {
        Ajax.request({
            url: ActionURL.buildURL('query', 'exportTabsXlsx.api'),
            method: 'POST',
            jsonData: {
                filename,
                tabModels,
            },
            downloadFile: true,
            success: () => {
                resolve();
            },
            failure: Utils.getCallbackWrapper(error => {
                console.error('Failed to export tabular data', error);
                reject(resolveErrorMessage(error) ?? 'Unexpected error while exporting selected tabs.');
            }),
        });
    });
}
```

labkey-nicka reviewed (changes requested):
A few things worth considering.

Sigmonia committed:
Update export action name and properties

Sigmonia committed:
v2.160.0-fb-excelTabs.6

Sigmonia committed:
Some lint/style clean-up
Adjust error handling for new action

Sigmonia committed:
Use api that allows reading error message

Sigmonia committed:
v2.160.0-fb-excelTabs.7

labkey-nicka commented on packages/components/src/internal/components/samples/SampleAssayDetail.tsx:
```diff
@@ -284,6 +284,9 @@ export const SampleAssayDetailBodyImpl: FC<SampleAssayDetailBodyProps & Injected
         );
     }
 
+    const row = sampleModel.getRow();
```
nit: Presuming `sampleModel` is a `QueryModel` you can do this instead:

```tsx
const sampleName = sampleModel.getRowValue('Name');
```

Sigmonia committed:
Add label to match grid

Sigmonia committed:
Adjust header value for SampleID -> Sample ID

Sigmonia committed:
merge from develop

Sigmonia committed:
address:
Issue #45388: Workflows excel export
Issue #45391: Excel export filename does not match table name as expected

Sigmonia committed:
v2.164.1-fb-excelTabs.8

Sigmonia committed:
add an export prefix string for SourceSamples/Assays

Sigmonia committed:
v2.164.1-fb-excelTabs.9

Sigmonia committed:
bump api version

Sigmonia committed:
Lint

Sigmonia committed:
missed yarn.lock

Sigmonia committed:
Finalize release notes and version number
